Add the release note for 1.16.0 (#4218)
line/armeria: site/src/pages/release-notes/1.16.0.mdx

---
date: 2022-04-19
---

## 🌟 New features

- Armeria <type://Server> now exposes the metrics about TLS handshake results. #4191
  ```
  armeria_server_tls_handshakes_total{
    cipher_suite=<the negotiated TLS cipher suite>,
    common_name=<the common name in TLS certificate>,
    protocol=<h1 or h2>,
    result=<success or failure>,
    tls_protocol=<TLS protocol name, e.g. TLSv1.3>
  }
  ```
- You can use the <type://@Delimiter> annotation in an annotated service to specify the delimiter of
  a query parameter or header. #4060
  ```java

>   public class MyAnnotatedService

Reviewer: ```suggestion
  public class MyAnnotatedService {
```

    // Given the request:
    //   GET /api/v1/users?ids=1,2,3
    // Armeria will inject:
    //   List.of(1, 2, 3)
    // into:
    //   List<Integer> ids
    @Get("/api/v1/users")
    public List<User> getUsers(@Param @Delimiter(",") List<Integer> ids) {
      ...
    }
  }
  ```
  - You can also specify a delimiter programmatically with <type://AnnotatedServiceBindingBuilder#queryDelimiter(String)>
- You can now use the no-op `subscribe()` to drain elements from a <type://StreamMessage>  #4145 #4185
  ```java
  // The following statement prints 1, 2 and 3.
  StreamMessage
    .of(1, 2, 3)
    .peek(value -> System.out.println(value))

>     .subscribe(); // 👈👈👈

Reviewer: Need `join()`?
```suggestion
    .subscribe().join(); // 👈👈👈
```

Author: That's a good idea :+1:

    }
    
    message GetNumbersRequest {}
    message GetNumbersResponse {
      repeated string my_number_array = 1; // 👈👈👈
    }
    ```
- You can now use the <type://@Decorator> annotation to apply decorators to your gRPC stubs. #3967 #4041
  ```java
  @LoggingDecorator // 👈👈👈
  public class GreeterImpl extends GreeterGrpc.GreeterImplBase {
    @Override
    @RateLimitingDecorator(/* requestsPerSec */ 10.0) // 👈👈👈 
    public void sayHello(
      HelloRequest req,
      StreamObserver<HelloReply> responseObserver
    ) {
      responseObserver.onNext(
        HelloReply
          .newBuilder()
          .setMessage(req.getName())
          .build()
      );
      responseObserver.onCompleted();
    }
  }
  ```
- <type://AbstractUnaryGrpcService> and unframed <type://GrpcService> now include the stack trace for failed
  requests when `verboseResponses` option is enabled for easier debuggability. #4114 #4171
- You can now specify query parameters additively by calling <type://WebClientRequestPreparation#queryParam(String,Object)>
  #4173
  ```java
  // Send GET /api/v1/greet?foo=1&bar=2
  WebClient
    .of().prepare()
    .get("/api/v1/greet")
    .queryParam("foo", "1")

>     .queryParams("bar", "2") // 👈👈👈

Reviewer: ```suggestion
    .queryParams(Map.of("bar", "2")) // 👈👈👈
```

      ServiceRequestContext ctx = GraphqlServiceContexts.get(env);
      ...
    }
  }
  ```
- You can now disable the warning message about masked routes when building a <type://Server> by specifying
  the `-Dcom.linecorp.armeria.reportMaskedRoutes=false` JVM option. #4086

## 📈 Improvements
 
- `armeria-graphql` doesn't use deprecated `graphql-java` API anymore. #4201
- Less verbose and clearer TLS handshake logging #4181

## 🛠️ Bug fixes

- JSON-gRPC transcoding now works even if you bind a <type://GrpcService> with path prefix. #4086 #4215
  ```java
  Server
    .builder()
    .route()
      .pathPrefix("/api") // 👈👈👈
      .build(
        GrpcService
          .builder()
          .enableHttpJsonTranscoding(true)
          .addService(...)
          .build()
      )
    .build()
  ```
- You now get the `HeaderListSizeException` as expected when the length of request headers is too long. #4180
- <type://EurekaEndpointGroup> now continues refreshing itself even if it failed to refresh once. #4206
- <type://HealthCheckedEndpointGroup> now leaves a detailed error message if all endpoints are unhealthy
  on the first check. #4200

> - Fixed a rare infinite loop in <type://WeightRampingUpStrategy> #TBD

Reviewer: ```suggestion
- Fixed a rare infinite loop in <type://EndpointSelectionStrategy#rampingUp()> #4219
```

Author: Thanks!
